### PR TITLE
[BIC-1903] Add configuration for is null / is not valued

### PR DESCRIPTION
**Pull request by JMGDoctolib**
Add a property `properties_always_exists_in_document` on model to be able to use `IS NULL` instead of `IS NOT VALUED` when `true`

### CODE REVIEW

File: lib/couchbase-orm/utilities/properties_always_exists_in_document.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+    module PropertiesAlwaysExistsInDocument
+        def properties_always_exists_in_document=(value)
+            unless [true, false].include? value
+                raise ArgumentError.new("properties_always_exists_in_document must be a boolean")
+            end
+            @properties_always_exists_in_document = value
+        end
+
+        def properties_always_exists_in_document
+            @properties_always_exists_in_document ||= false
```
Comment (pimpin) [inline lib/couchbase-orm/utilities/properties_always_exists_in_document.rb]: We may have highlighted that false is the default value using a const.

File: spec/base_spec.rb [hunk 2]
```diff
@@ -325,4 +328,16 @@ class InvalidNested < CouchbaseOrm::NestedDocument
             end
         end
     end
+
+    describe '.properties_always_exists_in_document' do
+        fit 'Uses NOT VALUED when properties_always_exists_in_document = false' do
```
Comment (pimpin) [inline spec/base_spec.rb]: What is that `fit` method ?
```suggestion
        it 'Uses NOT VALUED when properties_always_exists_in_document = false' do
```

File: spec/base_spec.rb [hunk 3]
```diff
@@ -325,4 +328,16 @@ class InvalidNested < CouchbaseOrm::NestedDocument
...
+            where_clause = BaseTest.where(name: nil)
+            expect(where_clause.to_n1ql).to eq("select raw meta().id from `billeo-cb-bucket` where type = 'base_test' AND name IS NOT VALUED order by meta().id ")
+        end
+
+        it 'Uses NOT VALUED when properties_always_exists_in_document = true' do
```
Comment (pimpin) [inline spec/base_spec.rb]:
```suggestion
        it 'Uses IS NULL when properties_always_exists_in_document = true' do
```

File: spec/base_spec.rb
Comment (pimpin) [inline spec/base_spec.rb:1]: Maybe we could also specify whats happen when properties_always_exist_in_document is set and a property is missing.

**Review by pimpin — APPROVED** [review]
LGTM

File: spec/base_spec.rb (hunk 2 above)
Comment (JMGDoctolib) [reply spec/base_spec.rb]: it's too only run this test but it's not working with this framework 😢

File: spec/base_spec.rb
Comment (JMGDoctolib) [reply spec/base_spec.rb:1]: Yes, by a comment you mean in the code + readme?
If I understand well it will crash, right?

**Review by pimpin — APPROVED** [review]

File: lib/couchbase-orm/utilities/query_helper.rb [hunk 4]
```diff
@@ -5,9 +5,12 @@ module QueryHelper
         module ClassMethods
 
             def build_match(key, value)
+                use_is_null = self.properties_always_exists_in_document
                 key = "meta().id" if key.to_s == "id"
                 case
-                when value.nil?
+                when value.nil? && use_is_null
+                    "#{key} IS NULL"
```
Comment (pimpin) [inline lib/couchbase-orm/utilities/query_helper.rb:12]: NB : This is should be `"#{key} IS NOT NULL"` (fixed in #35 )

**Pull request merged**